core/jvm/src/main/scala/scalaz/zio/blocking/Blocking.scala

```diff
@@ -55,23 +55,29 @@ object Blocking extends Serializable {
     def interruptible[A](effect: => A): ZIO[R, Throwable, A] =
       ZIO.flatten(ZIO.effectTotal {
         import java.util.concurrent.locks.ReentrantLock
-        import java.util.concurrent.atomic.AtomicReference
+        import java.util.concurrent.atomic.{ AtomicBoolean, AtomicReference }
         import scalaz.zio.internal.OneShot
 
-        val lock    = new ReentrantLock()
-        val thread  = new AtomicReference[Option[Thread]](None)
-        val barrier = OneShot.make[Unit]
+        val lock        = new ReentrantLock()
+        val interrupted = new AtomicBoolean(false)
+        val thread      = new AtomicReference[Option[Thread]](None)
+        val barrier     = OneShot.make[Unit]
 
         def withMutex[B](b: => B): B =
           try {
             lock.lock(); b
           } finally lock.unlock()
 
         val interruptThread: UIO[Unit] =
-          ZIO.effectTotal(withMutex(thread.get match {
-            case None         => ()
-            case Some(thread) => thread.interrupt()
-          }))
+          ZIO.effectTotal {
+            var isNone = false
+            while (!isNone && !withMutex(interrupted.get)) {
```

Comment:
Should this use a backoff?

Reply:
Yeah that while loop is scary. I first used a `Schedule` but `ensuring` won't accept the `R with Clock` environment and I don't have time to look into it right now.

Reply:
![image](https://user-images.githubusercontent.com/283332/54277531-cd1baf80-4598-11e9-88b5-e469d2ede982.png)